src/filesystem.c

```diff
@@ -275,6 +277,91 @@ rcutils_get_file_size(const char * file_path)
   return rc == 0 ? (size_t)(stat_buffer.st_size) : 0;
 }
 
+rcutils_ret_t
+rcutils_list_directory(
+  const char * directory_path,
+  rcutils_allocator_t allocator,
+  rcutils_string_array_t * string_array)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    directory_path, "directory_path is null", return RCUTILS_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    string_array, "string_array is null", return RCUTILS_RET_INVALID_ARGUMENT);
+
+  // Start with 8 entries
+  rcutils_ret_t ret = rcutils_string_array_init(string_array, 8, &allocator);
+  if (RCUTILS_RET_OK != ret) {
+    return ret;
+  }
+
+  size_t count = 0;
+
+#ifdef _WIN32
+  char * path = rcutils_join_path(directory_path, "*", allocator);
+  WIN32_FIND_DATA data;
+  HANDLE handle = FindFirstFile(path, &data);
+  allocator.deallocate(path, allocator.state);
+  if (INVALID_HANDLE_VALUE == handle) {
+    fprintf(stderr, "Can't open directory %s. Error code: %lu\n", directory_path, GetLastError());
+    ret = RCUTILS_RET_ERROR;
+    goto fail;
+  }
+  if (handle != INVALID_HANDLE_VALUE) {
+    do {
+      if (count >= string_array->size) {
+        ret = rcutils_string_array_resize(string_array, count * 2);
+        if (RCUTILS_RET_OK != ret) {
+          goto fail;
+        }
+      }
+
+      string_array->data[count] = rcutils_strdup(data.cFileName, allocator);
+      if (NULL == string_array->data[count]) {
+        goto fail;
+      }
+    } while (++count, FindNextFile(handle, &data));
+    FindClose(handle);
+  }
+#else
+  DIR * dir = opendir(directory_path);
+  if (NULL == dir) {
+    fprintf(stderr, "Can't open directory %s. Error code: %d\n", directory_path, errno);
+    ret = RCUTILS_RET_ERROR;
+    goto fail;
+  }
+  struct dirent * entry;
+  for (; NULL != (entry = readdir(dir)); ++count) {
```

Comment:
counting `.` and `..` directory included, is this intended?

Reply:
Yeah, I felt that we can leave it up to the consumer whether they want to ignore those entries rather than performing the comparisons here. The API proposed in #323 mentions this behavior in the API docs.